src/main/kotlin/build/buf/gradle/AbstractBufTask.kt

```diff
@@ -0,0 +1,32 @@
+// Copyright 2024 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package build.buf.gradle
+
+import org.gradle.api.DefaultTask
+import org.gradle.api.provider.Property
+import org.gradle.api.tasks.Input
+import java.io.File
+import java.nio.file.Path
+
+abstract class AbstractBufTask : DefaultTask() {
+    // this is just an @Input, not @InputDirectory
+    // this property is used only for relativization of paths, the task doesn't depend on the content
+    @get:Input
+    abstract val projectDir: Property<File>
```

Comment:
Is there any way to restrict the visibility of all of these properties on the various task classes? I'd rather them not become part of the plugin's public API.

Also, while it's fresh, I think we should get proper KDoc on each property (even if we can make them internal). I'll never remember what they are a year from now...

Reply:
Added a few kdocs, although their usefulness is a bit low.
Not sure about the way to hide the inputs/outputs - kotlin internal doesn't work, Internal annotation in gradle has different meaning and semantics.

Reply:
There's an `@Internal` Gradle annotation - I wonder what that does.

E.g. https://github.com/google/protobuf-gradle-plugin/blob/master/src/main/groovy/com/google/protobuf/gradle/ProtobufExtract.groovy#L71

Reply:
> although their usefulness is a bit low.

I happen to think the nontrivial ones (and even some of the simpler ones) will be helpful and establish good practice. Thanks for adding them!

Reply:
For gradle `@Internal`, see the docs: https://docs.gradle.org/current/javadoc/org/gradle/api/tasks/Internal.html
In particular, "This will cause the task not to be considered out-of-date when the property has changed.", which is not what we want.

Reply:
The issue with kotlin `internal` is that I cannot put that on parent classes (`AbstractBufExecTask`, `AbstractBufTask`) - I don't know exactly why. Putting it just on the concrete classes seems very inconsistent.

Reply:
When I change `AbstractBufTask.projectDir` to internal, the test fail on
```
FAILURE: Build failed with an exception.

* What went wrong:
Could not determine the dependencies of task ':publishBufImagePublicationPublicationToMavenRepository'.
> Could not create task ':bufBuild'.
   > 'org.gradle.api.provider.Property build.buf.gradle.AbstractBufTask.getProjectDir()'
```